src/IIIFPresentation/API/Features/Storage/Requests/PostHierarchicalCollection.cs

```diff
@@ -0,0 +1,130 @@
+﻿using System.Data;
+using API.Auth;
+using API.Converters;
+using API.Features.Storage.Helpers;
+using API.Helpers;
+using API.Infrastructure.Requests;
+using API.Settings;
+using AWS.S3;
+using AWS.S3.Models;
+using Core;
+using IIIF.Presentation.V3;
+using IIIF.Presentation.V3.Content;
+using IIIF.Serialisation;
+using MediatR;
+using Microsoft.Extensions.Options;
+using Models.API.General;
+using Repository;
+using Repository.Helpers;
+using DatabaseCollection = Models.Database.Collections;
+using IIdGenerator = API.Infrastructure.IdGenerator.IIdGenerator;
+
+namespace API.Features.Storage.Requests;
+
+public class PostHierarchicalCollection(
+    int customerId,
+    string slug, 
+    UrlRoots urlRoots,
+    string rawRequestBody) : IRequest<ModifyEntityResult<Collection, ModifyCollectionType>>
+{
+    public int CustomerId { get; } = customerId;
+
+    public string Slug { get; } = slug;
+    
+    public UrlRoots UrlRoots { get; } = urlRoots;
+    
+    public string RawRequestBody { get; } = rawRequestBody;
+}
+
+public class PostHierarchicalCollectionHandler(
+    PresentationContext dbContext,    
+    ILogger<CreateCollection> logger,
+    IBucketWriter bucketWriter,
+    IIdGenerator idGenerator,
+    IOptions<ApiSettings> options)
+    : IRequestHandler<PostHierarchicalCollection, ModifyEntityResult<Collection, ModifyCollectionType>>
+{
+    private readonly ApiSettings settings = options.Value;
+    
+    public async Task<ModifyEntityResult<Collection, ModifyCollectionType>> Handle(PostHierarchicalCollection request,
+        CancellationToken cancellationToken)
+    {
+        var splitSlug = request.Slug.Split('/');
+
+        var parentSlug = String.Join(String.Empty, splitSlug.Take(..^1));
+        var parentCollection =
+            await dbContext.RetriveHierarchicalCollection(request.CustomerId, parentSlug, cancellationToken);
+        
+        if (parentCollection == null) return ErrorHelper.NullParentResponse<Collection>();
+
+        string id;
+        
+        try
+        {
+            id = await dbContext.Collections.GenerateUniqueIdAsync(request.CustomerId, idGenerator, cancellationToken);
+        }
+        catch (ConstraintException ex)
+        {
+            logger.LogError(ex, "An exception occured while generating a unique id");
```

Comment:
Do these logs contain a correlation_id or anything to help know more context around the request?

Reply:
it currently just contains the standard exception information - I'll look to add the `CorrelationIdHeaderEnricher` from protagonist

Reply:
That can be done separately I think to save adding more to this. It's definitely useful to add as we can track requests across iiif-p and protagonist-api/engine etc. Create a ticket and we can pick it up

Reply:
https://github.com/dlcs/iiif-presentation/issues/70